01/implement-pascals.checkpoint.md

```diff
@@ -0,0 +1,202 @@
+
+The function will return a 2D array where the outer list stores the entire triangle and each index of the list stores the numbers in row `index + 1` of Pascal's triangle. For example, index 0 will store the first row, index 1 will store the second row, and so on. 
+
+The first row will always contain [1], the second row will always contain [1, 1]. Each row afterwards will have 1 as the first and last element and the middle elements will be calculated using the two numbers directly above it.
+
+If the function receives a number that is less than 0, a `ValueError` is thrown.
+
+(This problem is sourced from [LeetCode](https://leetcode.com/problems/pascals-triangle/description/))
+
+**Example 1:**
+
+```
+Input: num_rows = 3
+Output: [[1], [1, 1], [1, 2, 1]]
+Explanation: The 0th index (first row) in the outer list contains one number: [1]. 
+The 1st index (second row) contains two numbers: [1, 1]. 
+The 2nd index (third row) can be calculated by setting the first and last indexes to 1 
+and calculating the middle index using the two elements above it (1 + 1) 
+so the third row is [1, 2, 1].
+```
+
+**Example 2:**
+```
+Input: num_rows = 0
+Output: []
+Explanation: If there are 0 rows, the outer list will be empty.
+```
+
+**Example 3:**
+```
+Input: num_rows = 6
+Output: [
+    [1], 
+    [1, 1], 
+    [1, 2, 1], 
+    [1, 3, 3, 1], 
+    [1, 4, 6, 4, 1], 
+    [1, 5, 10, 10, 5, 1]
+]
+Explanation: The first index (first row) in the outer list contains one number: [1].
+The second index (second row) contains two numbers: [1, 1]. 
+Each row afterwards can be calculated by setting the first and last index to 1 
+and calculating the middle elements using the two elements above it.
+```
+
+### !end-question
+### !placeholder
+
+```python
+def pascals_triangle(num_rows):
+    """
+    A function that returns the first num_rows of Pascal's triangle.
+  
+    Parameters:
+    num_rows (int): The number of rows to return
+  
+    Returns:
+    list[list[int]]: 2D array where the outer list stores the entire triangle 
+    and each index of the list stores the numbers in row `index + 1` 
+    of Pascal's triangle.
+    """
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_pascals_nominal(self):
+        # Arrange
+        numRows = 3
+
+        # Act
+        result = pascals_triangle(numRows)
+
+        # Assert
+        self.assertEqual(result, [[1], [1, 1], [1, 2, 1]])
+
+    def test_pascals_big_number(self):
+        # Arrange
+        numRows = 10
+
+        # Act
+        result = pascals_triangle(numRows)
+
+        # Act
+        self.assertEqual(result, [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1], [1, 5, 10, 10, 5, 1], [1, 6, 15, 20, 15, 6, 1], [1, 7, 21, 35, 35, 21, 7, 1], [1, 8, 28, 56, 70, 56, 28, 8, 1], [1, 9, 36, 84, 126, 126, 84, 36, 9, 1]])
+
+    def test_pascals_base_case(self):
```

Comment:
Consider adding a case to test the other base case `numRows = 1`

Reply:
Added 👍🏾